### PR TITLE
PIL-1035: Feature/multi chain native asset send

### PR DESCRIPTION
**Pull request by poocart**
https://linear.app/pillarproject/issue/PIL-1035/send-flows-side-chain-integration

This includes flow for sending multi chain native asset and collectibles. Multi chain tokens can be sent using these changes, but there's no feature yet to pull multi chain tokens.

Big in size because it adds additional params to few methods that are shared cross codebase.

This does not cover and will add separate PR to this:
- multi chain history, this is not ready per back-end, but will cover locally saved multi chain history items.
- multi chain supported tokens (can be sent, but not being pulled).

Note: per yesterday's V2 catch up call I'm not sure if `chainId` missing Wallet Connect should block user from going further or there should always be a fallback to Ethereum mainnet?

History as separate PRs to be included before merge:
- [x] Multi chain history (https://github.com/pillarwallet/pillarwallet/pull/2694)

Supported assets will be added later.

### CODE REVIEW

File: src/actions/walletConnectActions.js [hunk 1]
```diff
@@ -362,6 +362,13 @@ export const estimateWalletConnectCallRequestTransactionAction = (callRequest: W
 
     const { amount: value, to, data } = mapCallRequestToTransactionPayload(callRequest, accountAssets, supportedAssets);
 
-    dispatch(estimateTransactionAction({ value, to, data }));
+    const { chainId } = callRequest;
+    const chain = chainFromChainId[chainId];
+    if (!chain) {
+      dispatch(setWalletConnectErrorAction(t('error.walletConnect.cannotDetermineEthereumChain')));
```
Comment (IAmKio) [inline src/actions/walletConnectActions.js:368]: 👍

File: src/utils/transactions.js [hunk 2]
```diff
@@ -79,12 +82,14 @@ export const buildEthereumTransaction = async (
...
   tokenId: ?string,
+  chain: Chain,
 ): Promise<EthereumTransaction> => {
   let value;
 
   if (tokenType !== COLLECTIBLES) {
+    const chainNativeSymbol = nativeSymbolPerChain[chain];
     value = utils.parseUnits(amount, decimals);
-    if (symbol !== ETH && !data && contractAddress) {
+    if (symbol !== chainNativeSymbol && !data && contractAddress) {
```
Comment (mdjastrzebski) [inline src/utils/transactions.js:92]: 👍🏻

File: src/utils/history.js [hunk 3]
```diff
@@ -196,3 +202,87 @@ export const getTokenTransactionsFromHistory = (
...
+      gasToken: { decimals, symbol },
+    } = feeWithGasToken;
+    return {
+      value: wrapBigNumber(formatUnits(feeInWei, decimals)),
+      symbol,
+    };
+  }
+
+  if (gasUsed && gasPrice) {
+    const feeValue = wrapBigNumber(gasPrice).multipliedBy(wrapBigNumber(gasPrice));
```
Comment (mdjastrzebski) [inline src/utils/history.js:223]: BN operations arguments seem to work fine with `number | string` too
```suggestion
    const feeValue = wrapBigNumber(gasPrice).multipliedBy(gasPrice);
```

File: src/services/etherspot.js [hunk 4]
```diff
@@ -278,16 +278,36 @@ export class EtherspotService {
...
+      reportErrorLog('clearTransactionsBatch failed: no SDK for chain set', { chain });
+      throw new Error(t('error.unableToResetTransactions'));
+    }
+
+    sdk.clearGatewayBatch();
   }
 
-  setTransactionsBatch(transactions: EthereumTransaction[]) {
-    return Promise.all(transactions.map((transaction) => this.sdk.batchExecuteAccountTransaction(transaction)));
+  setTransactionsBatch(transactions: EthereumTransaction[], chain: Chain) {
```
Comment (mdjastrzebski) [inline src/services/etherspot.js:292]: Consider moving `chain` param in etherspotService methods to the front, as this is kind of scope for the whole operation

```suggestion
  setTransactionsBatch(chain: Chain, transactions: EthereumTransaction[]) {
```

File: src/screens/SendToken/SendTokenConfirm.js [hunk 5]
```diff
@@ -19,108 +19,96 @@
...
-  session: Object,
-};
+const SendTokenConfirm = () => {
+  const session = useRootSelector(({ session: sessionState }) => sessionState.data);
+  const navigation = useNavigation();
 
-class SendTokenConfirm extends React.Component<Props> {
-  source: string;
+  const source = useNavigationParam('source');
+  const transactionPayload = useNavigationParam('transactionPayload');
```
Comment (mdjastrzebski) [inline src/screens/SendToken/SendTokenConfirm.js:49]: Pls add explicit typing to these params, consider adding some fallback code if they come null/undefined

```suggestion
  const source = useNavigationParam('source');
  const transactionPayload = useNavigationParam('transactionPayload');
```

File: src/screens/SendCollectible/SendCollectibleConfirm.js [hunk 6]
```diff
@@ -88,19 +89,22 @@ const WarningMessage = styled(Paragraph)`
...
   isEstimating,
   estimateErrorMessage,
 }: Props) => {
+  const navigation = useNavigation();
+
+  const receiverEnsName = useNavigationParam('receiverEnsName');
+  const assetData = useNavigationParam('assetData');
+  const receiver = useNavigationParam('receiver');
+  const navigationSource = useNavigationParam('source');
+  const chain = useNavigationParam('chain');
```
Comment (mdjastrzebski) [inline src/screens/SendCollectible/SendCollectibleConfirm.js:105]: Pls add explicit typing to these params, consider adding fallback code if any if the params is null/underifned

```suggestion
  const receiverEnsName = useNavigationParam('receiverEnsName');
  const assetData = useNavigationParam('assetData');
  const receiver = useNavigationParam('receiver');
  const navigationSource = useNavigationParam('source');
  const chain = useNavigationParam('chain');
```

File: src/screens/Asset/Asset.js [hunk 7]
```diff
@@ -155,207 +166,204 @@ const SyntheticAssetIcon = styled(Image)`
...
-    const { navigation } = this.props;
-    const { assetData } = navigation.state.params;
-    this.props.navigation.navigate(SEND_TOKEN_FROM_ASSET_FLOW, { assetData });
-  };
-
-  goToExchangeFlow = (fromAssetCode: ?string, toAssetCode?: string) => {
-    this.props.navigation.navigate(EXCHANGE, { fromAssetCode, toAssetCode });
+  }, []);
+
+  const assetData = useNavigationParam('assetData');
```
Comment (mdjastrzebski) [inline src/screens/Asset/Asset.js:194]: PLs add explicit typing for nav params

File: src/screens/Asset/Asset.js [hunk 8]
```diff
@@ -155,207 +166,204 @@ const SyntheticAssetIcon = styled(Image)`
...
+            {isArchanovaAccount(activeAccount) && (
+              <ActivityFeed
+                feedTitle={t('title.transactions')}
+                navigation={navigation}
+                feedData={transactions}
+                isAssetView
+              />
+            )}
+            {/* $FlowFixMe: should be fine after Archanova history mappings are discarded */}
+            {isEtherspotAccount(activeAccount) && <HistoryList items={transactions} />}
```
Comment (mdjastrzebski) [inline src/screens/Asset/Asset.js:354]: 👍🏻 
There is a similar history section in collectible details, could you add etherspot history list data there as well?

**Review by mdjastrzebski — APPROVED** [review]
Looks great man! 🚀 🌕 

Added some stylistic/edge-case comments but didn't spot any major issues.

File: src/containers/SendContainer/SendContainer.js [hunk 9]
```diff
@@ -112,9 +114,16 @@ const SendContainer = (props: Props) => {
     customScreenTitle,
   } = props;
 
+  const chain = customValueSelectorProps?.assetData?.chain ?? CHAIN.ETHEREUM;
+  const { title: chainTitle } = useChainsConfig()[chain];
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/containers/SendContainer/SendContainer.js:118]: It's long one per design.

File: src/screens/Asset/Asset.js (hunk 8 above)
Comment (poocart) [reply src/screens/Asset/Asset.js:354]: Will add include into multi chain history PR.

**Review by IAmKio — APPROVED** [review]
Great work mate, can't wait to see this in action! 🚀

File: src/components/SendAsset/SendAsset.js [hunk 10]
```diff
@@ -240,11 +247,12 @@ const SendAsset = ({
...
     if (percentageModifier === 1.00 && selectedContact) {
       // await needed for initial max available send calculation to get estimate before showing max available after fees
-      await estimateTransaction({
+      const transactionToEstimate = {
         to: selectedContact.ethAddress,
         value: calculatedBalanceAmount.toString(),
         assetData: mapToAssetDataType(assetData),
-      });
+      };
+      await estimateTransaction(transactionToEstimate, chain);
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/components/SendAsset/SendAsset.js:255]: Nothing, this dispatches action that returns estimate results from reducer as well as `Toas`t is shown if something goes wrong, that is sitting under single action.

File: src/actions/transactionEstimateActions.js [hunk 11]
```diff
@@ -148,18 +153,18 @@ export const estimateTransactionsAction = (transactionsToEstimate: TransactionTo
...
         }
 
-        await etherspotService.setTransactionsBatch(transactions).catch((error) => {
+        await etherspotService.setTransactionsBatch(chain, transactions).catch((error) => {
           errorMessage = error?.message;
         });
 
-        estimated = await etherspotService.estimateTransactionsBatch(gasToken?.address).catch((error) => {
+        estimated = await etherspotService.estimateTransactionsBatch(gasToken?.address, chain).catch((error) => {
           errorMessage = error?.message;
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/transactionEstimateActions.js:168]: Yes, we're capturing inside the SDK `estimateTransactionsBatch` method.

File: src/components/ValueInput/ValueInput.js [hunk 12]
```diff
@@ -160,7 +161,9 @@ export const ValueInputComponent = ({
   const ratesWithCustomRates = { ...rates, ...customRates };
 
   const assetSymbol = assetData.symbol || '';
-  const assetBalance = (customBalances || balances)[assetSymbol]?.balance || '0';
+  const chain = assetData?.chain || CHAIN.ETHEREUM;
+  const walletBalances = accountAssetsBalances?.[chain]?.wallet ?? {};
+  const assetBalance = (customBalances || walletBalances)[assetSymbol]?.balance || '0';
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/components/ValueInput/ValueInput.js:166]: None of the methods below that use `assetBalance` expects `BigNumber` so in this case I would leave as it is. Changes were not added on this line so nothing should break.

**Pull request merged**